fix(remotecluster): can not upgrade helm release

nioshield opened the pull request:
<!--
Thank you for contributing to Chaos Mesh!

If you're unsure where to start, please refer to the contributing doc:

https://github.com/chaos-mesh/chaos-mesh/blob/master/CONTRIBUTING.md

If you still have questions, please let us know via issues.

Please follow https://www.conventionalcommits.org/en/v1.0.0/ when you open a new PR:
-->

## What problem does this PR solve?

this pr is part of #4063


## What's changed and how it works?
* add `RemoteClusterStatus.ObservedGeneration`  config and save current remotecluster version
* check helm install and upgrade
* fix `RemoteClusterSpec.ConfigOverride` decode value err

## Related changes

- [ ] This change also requires further updates to the [website](https://github.com/chaos-mesh/website) (e.g. docs)
- [ ] This change also requires further updates to the `UI interface`

## Cherry-pick to release branches (optional)

> This PR should be cherry-picked to the following release branches:

- [x] release-2.6
- [ ] release-2.5

## Checklist

### CHANGELOG

> Must include at least one of them.

- [ ] I have updated the `CHANGELOG.md`
- [ ] I have labeled this PR with "no-need-update-changelog"

### Tests

> Must include at least one of them.

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test

### Side effects

- [ ] **Breaking backward compatibility**

## DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

nioshield commented on pkg/helm/helm_client.go:
```diff
@@ -96,6 +96,20 @@ func (h *HelmClient) UpgradeOrInstall(namespace string, releaseName string, char
 	return result, nil
 }
 
+func (h *HelmClient) UpgradeRelease(namespace string, releaseName string, chart *chart.Chart, values map[string]interface{}) (*release.Release, error) {
```
updated  test case for this file , please take a look.

g1eny0ung commented on controllers/multicluster/remotecluster/controller.go:
```diff
@@ -189,26 +201,36 @@ func (r *Reconciler) ensureHelmRelease(ctx context.Context, obj *v1alpha1.Remote
+	values := make(map[string]interface{})
+	if obj.Spec.ConfigOverride != nil {
+		configStr := strings.Trim(string(obj.Spec.ConfigOverride), "\"")
+		configByte, err := base64.StdEncoding.DecodeString(configStr)
```
The type of `configOverride` is `json.RawMessage`, not `base64`. I think you're using the one that suits your situation here, but it causes incompatibility. You should revert this change.

nioshield commented:
@g1eny0ung Please take a look

nioshield commented:
@g1eny0ung Any other suggestions?

g1eny0ung commented:
> @g1eny0ung Any other suggestions?

@nioshield I still not review the recent changes, so busy last week. 🫨 I will continue to deal with this week.

STRRL commented:
It seems use base64 encoded string in configOverride not works...

this yaml does not work for me

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
 name: remote-cls
spec:
 namespace: "chaos-mesh"
 version: "2.5.0"
 configOverride: eyJkYXNoYm9hcmQiOnsiY3JlYXRlIjpmYWxzZX19    //{"dashboard":{"create":false}}
 kubeConfig:
   secretRef:
      name: remote-cls
      namespace: default
      key: remote-cls
```

What should I set in field `configOverride` 🤔?

nioshield commented:
Using base64-encoded strings in configOverride requires special handling of decoded strings, using an unstructured implementation of configOverride, which I'm working on in another branch

STRRL commented:
> Using base64-encoded strings in configOverride requires special handling of decoded strings, using an unstructured implementation of configOverride, which I'm working on in another branch

Could I use `configOverride` to configure the RemoteCluster with proper value? Or I have to wait the the work.

g1eny0ung commented:
After referring to https://github.com/kubernetes-sigs/controller-tools/issues/637, I added the appropriate markers and updated `controller-gen` to `v0.6.1`. Following this, I regenerated the CRDs by running `make manifests/crd.yaml`.

After experimenting, I added values to `configOverride` and reapplied. Based on the results, it appears that `json.Unmarshal` is properly handling the changes.

An example:

```yaml
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
  name: remote-cluster-test
spec:
  namespace: "chaos-mesh"
  version: "2.6.1"
  configOverride:
    dashboard:
      create: false
  kubeConfig:
    secretRef:
      name: secret-kubeconfig
      namespace: default
      key: kubeconfig
```

@nioshield @STRRL Can you help verify this when you have time?

nioshield commented:
> After referring to [kubernetes-sigs/controller-tools#637](https://github.com/kubernetes-sigs/controller-tools/issues/637), I added the appropriate markers and updated `controller-gen` to `v0.6.1`. Following this, I regenerated the CRDs by running `make manifests/crd.yaml`.
> 
> After experimenting, I added values to `configOverride` and reapplied. Based on the results, it appears that `json.Unmarshal` is properly handling the changes.
> 
> An example:
> 
> ```yaml
> apiVersion: chaos-mesh.org/v1alpha1
> kind: RemoteCluster
> metadata:
>   name: remote-cluster-test
> spec:
>   namespace: "chaos-mesh"
>   version: "2.6.1"
>   configOverride:
>     dashboard:
>       create: false
>   kubeConfig:
>     secretRef:
>       name: secret-kubeconfig
>       namespace: default
>       key: kubeconfig
> ```
> 
> @nioshield @STRRL Can you help verify this when you have time?
I'll check it out later

STRRL commented:
This PR could not fully resolve the #4063, and we need ongoing changes to fix `configOverride` field.


I would remove the close instruction in the pr's description, and I would review it again in soon

STRRL commented:
it seems it's required to change the type of `configOverride` to something like

https://github.com/fluxcd/helm-controller/blob/98a5a518a1e717bb379a4f1e7cfc594ea2c78653/api/v2beta1/helmrelease_types.go#L181

g1eny0ung commented:
> it seems it's required to change the type of `configOverride` to something like
> 
> https://github.com/fluxcd/helm-controller/blob/98a5a518a1e717bb379a4f1e7cfc594ea2c78653/api/v2beta1/helmrelease_types.go#L181-L183

It looks like this also handles `key/value` structures correctly. Have you tried it yourself? 🤔 But I still feel that using these markers is simple and do not need to rely on extra packages:

```go
// +kubebuilder:validation:Schemaless
// +kubebuilder:pruning:PreserveUnknownFields
// +kubebuilder:validation:Type=object
```

g1eny0ung commented:
> This PR could not fully resolve the #4063, and we need ongoing changes to fix `configOverride` field.
> 
> I would remove the close instruction in the pr's description, and I would review it again in soon

Can you give more details? Is there anything else we need to take care of?

nioshield commented:
> it seems it's required to change the type of `configOverride` to something like
> 
> https://github.com/fluxcd/helm-controller/blob/98a5a518a1e717bb379a4f1e7cfc594ea2c78653/api/v2beta1/helmrelease_types.go#L181-L183

ConfigOverride is modified in this way and works properly
An example:
```yml
apiVersion: chaos-mesh.org/v1alpha1
kind: RemoteCluster
metadata:
 name: remote-cls
spec:
 namespace: "chaos-mesh"
 version: "2.5.0"
 configOverride:
   dashboard:
      create: false
   controllerManager:
      replicaCount: 2
 kubeConfig:
   secretRef:
      name: remote-cls
      namespace: default
      key: remote-cls
```
@STRRL @g1eny0ung Please look at it again

g1eny0ung commented on CHANGELOG.md:
```diff
@@ -87,6 +87,7 @@ For more information and how-to, see [RFC: Keep A Changelog](https://github.com/
 - Fix version comparison in install.sh [#3901](https://github.com/chaos-mesh/chaos-mesh/pull/3901)
 - Fix stuck dashboard updates when using ReadWriteOnce PVCs [#3876](https://github.com/chaos-mesh/chaos-mesh/issues/3876)
 - Fix MySQL NO_ZERO_IN_DATE by using `*time.Time` to represent finish time [#4056](https://github.com/chaos-mesh/chaos-mesh/pull/4056)
+- Fix remote cluster cannot upgrade helm release [#4075](https://github.com/chaos-mesh/chaos-mesh/pull/4075)
```
Good job!

There is one remaining matter that requires attention. Kindly place this line in the unreleased section above. Additionally, there is no need to force-push the changes repeatedly as this results in the loss of context.

g1eny0ung commented:
@nioshield I tried testing the recent changes locally and the results were exactly what I was hoping for! 🥳

But there is a problem with the API server to validate the `configOverride` field if it is passed as a string or anything other than an object. This is due to the absence of an object type restriction.

So I still prefer the solution I found, append below markers and keep the `json.RawMessage` (which also works fine):

```go
// +kubebuilder:validation:Schemaless
// +kubebuilder:validation:Type=object
// +kubebuilder:pruning:PreserveUnknownFields
ConfigOverride json.RawMessage `json:"configOverride,omitempty"`
```

And the solution from @STRRL:

```go
ConfigOverride *apiextensionsv1.JSON `json:"configOverride,omitempty"`
```

I reviewed created CRDs and discovered that using `*apiextensionsv1.JSON` will add `x-kubernetes-preserve-unknown-fields: true` to the OpenAPI definitions. This is equivalent to appending the `// +kubebuilder:pruning:PreserveUnknownFields` marker. If we choose to use `*apiextensionsv1.JSON` as the type of `ConfigOverride`, we still need to add the validation markers described above. I think that would be a bit redundant.

nioshield commented:
@g1eny0ung Refix and use the markup
```
// +kubebuilder:validation:Schemaless
// +kubebuilder:validation:Type=object
// +kubebuilder:pruning:PreserveUnknownFields
ConfigOverride json.RawMessage `json:"configOverride,omitempty"`
```

g1eny0ung reviewed (approved):
/lgtm

Thanks so much for contributing these past two months! 👍🏼

g1eny0ung commented:
/cherrypick release-2.6
